lib/iris/experimental/stratify.py

```diff
@@ -0,0 +1,178 @@
+        dims = cube.coord_dims(coord)
+        if z_dim not in dims:
+            result.add_aux_coord(coord.copy(), dims)
+
+
+def relevel(phenom, level_data, target_levels, interp_direction=0,
+            interpolator=None):
+    """
+    Interpolate the given phenomenon onto the given levels of level_data.
+
+    For example, suppose we have two datasets `P(i,j,k)` and `H(i,j,k)`
+    and we want ``P(i,j,H)`` we call relevel with
+    ``phenom=P`` and ``level_data=H`` and ``target_levels`` being
+    an array of the values of ``H`` we would like.
+
+    This routine is especially useful for computing isosurfaces of
+    phenomenon that are generally monotonic in the direction of interpolation
+    (such as height/pressure and salinity/depth).
+
+    Parameters
+    ----------
+    phenom
+        The data to be re-levelled.
+    
+    level_data : iris.cube.Cube or iris.coord.Coord
+        The coordinate values (cube/coord) in the same system as
+        ``target levels``.
+        All dimensions of ``level_data`` must be broadcastable to ``phenom``
+
+    target_levels : array-like or iris.cube.Cube
+        The levels of ``level_data`` to pick from ``phenom``.
+        For the default interpolator, this must be a 1d array.
+        For a custom interpolator this may be multi-dimensional and must
+        broadcast to phenom and level_data except in the phenomenon's
+        ``interp_direction`` dimension.
+
+    interp_direction : int or :class:`~iris.coords.Coord` instance
+        The direction of interpolation. This is necessary because
+        level_data is rarely 1D. This is often the z dimension coordinate
+        of ``phenom``.
+
+    interpolator : callable or None
+        The interpolator to use when computing the interpolation. The function
+        will be passed the following positional arguments::
+
+            (target levels, level data, phenom data, interp axis)
+
+        If the interpolator is None, :func:`stratify.interpolate` will be used
+        with linear interpolation and NaN extrapolation.
+
+        An example of constructing an alternative interpolation scheme:
+
+            from functools import partial
+            interpolator = partial(stratify.interpolate,
+                                   interpolation=stratify.INTERPOLATE_NEAREST,
+                                   extrapolation=stratify.EXTRAPOLATE_LINEAR)
+
+    """
+    # Identify the z-coordinate within the phenom cube
+    if isinstance(interp_direction, Coord):
+        [interp_dim] = phenom.coord_dims(interp_direction)
+    else:
+        interp_dim = interp_direction
+
+    if isinstance(level_data, Coord):
+        source_data = level_data.points
+    else:
+        source_data = level_data.data 
+
+    # phenom and level_data must be broadcastable.
+    phenom_data, source_data = np.broadcast_arrays(phenom.data,
+                                                   source_data)
+
+    target_levels = np.array(target_levels)
```

Comment:
Sorry @pelson I'm finding difficult to follow, what equation?? :)

In any case, I may not have been clear from my original comment.  I meant, I don't see how passing a cube rather than an array is being dealt with by stratify.
Since you suggest that target_levels can be a cube, this line would convert it to an array with one element (a cube).  This should then fall over in stratify when it specifies a dtype of float64 should it not?? (https://github.com/SciTools-incubator/python-stratify/blob/master/stratify/_vinterp.pyx#L465)

I have not tested this, but neither have you as far as I can tell.